athena-federation-sdk/src/main/java/com/amazonaws/athena/connector/lambda/domain/predicate/FederationExpressionParser.java

```diff
@@ -0,0 +1,137 @@
+{
+    static final Logger LOGGER = LoggerFactory.getLogger(FederationExpressionParser.class);
+
+    private static final String quoteCharacter = "'"; // AFAIK this is going to be valid for all sources when quoting a constant.
+
+    /**
+     * Each datasource has different syntax for various operations, quotes, etc. This is the only method a subclass to implement, and otherwise will just
+     * invoke parseComplexExpressions.
+     */
+    public abstract String mapFunctionToDataSourceSyntax(FunctionName functionName, ArrowType type, List<String> arguments);
+
+    public List<String> parseComplexExpressions(List<Field> columns, Constraints constraints)
+    {
+        List<String> complexExpressionConjuncts = new ArrayList<>();
+        if (constraints.getExpression() == null || constraints.getExpression().isEmpty()) {
+            return complexExpressionConjuncts;
+        }
+
+        List<FederationExpression> federationExpressions = constraints.getExpression();
+        for (FederationExpression federationExpression : federationExpressions) {
+            FunctionCallExpression functionCallExpression = (FunctionCallExpression) federationExpression;
+            complexExpressionConjuncts.add(parseFunctionCallExpression(functionCallExpression));
+        }
+        return complexExpressionConjuncts;
+    }
+
+    /**
+     * This is a recursive function, as function calls can have arguments which, themselves, are function calls.
+     * @param functionCallExpression
+     * @return
+     */
+    public String parseFunctionCallExpression(FunctionCallExpression functionCallExpression)
+    {
+        return parseFunctionCallExpressionHelper(functionCallExpression);
+    }
+
+    protected String parseFunctionCallExpressionHelper(FunctionCallExpression functionCallExpression)
+    {
+        FunctionName functionName = functionCallExpression.getFunctionName();
+        List<FederationExpression> functionArguments = functionCallExpression.getArguments();
+        List<String> arguments = new ArrayList<>();
+
+        for (FederationExpression argument : functionArguments) {
+            if (argument instanceof FunctionCallExpression) { // recursive case
+                arguments.add(parseFunctionCallExpressionHelper((FunctionCallExpression) argument));
+            } 
+            else if (argument instanceof ConstantExpression) { // base case
+                ConstantExpression constantExpression = (ConstantExpression) argument;
+                arguments.add(parseConstantExpression(constantExpression));
+            }
+            else if (argument instanceof VariableExpression) { // base case
+                VariableExpression variableExpression = (VariableExpression) argument;
+                arguments.add(parseVariableExpression(variableExpression));
+            } 
+            else {
+                throw new RuntimeException("Should not reach this case - a new subclass was introduced and is not handled.");
+            }
```

Comment:
nit:

```
List<String> arguments = functionArguments.stream().map(argument -> {
    // base cases first
    if (argument instanceof ConstantExpression) {
        return parseConstantExpression((ConstantExpression) argument);
    }
    else if (argument instanceof VariableExpression) {
        return parseVariableExpression((VariableExpression) argument);
    }
    // recursive case
    else if (argument instanceof FunctionCallExpression) {
        return parseFunctionCallExpressionHelper((FunctionCallExpression) argument);
    }
    throw new RuntimeException("Should not reach this case - a new subclass was introduced and is not handled.");
}).collect(Collectors.toList());
```

Reply:
fixed in latest. thanks for the video, great watch